### PR TITLE
Add group matrix.

### PR DESCRIPTION
**Pull request by tsiegleauq**
Create, Read, Update, Delete works

### CODE REVIEW

**Comment by tsiegleauq** [comment]
@FinnStutzenstein please review

I will work on some general UI UX improvements tomorrow (which will also cover the group page)
so don't pay too much attention to the ugly mini-forms there ;-)

File: client/src/app/site/users/services/group-repository.service.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * Add an entry to appPermissions
+     *
+     * @param appId number that indicates the app
+     * @param perm certain permission as string
+     * @param appName Indicates the header in the Permission Matrix
+     */
+    private addAppPerm(appId: number, perm: string, appName: string): void {
+        if (!this.appPermissions[appId]) {
+            this.appPermissions['' + appId] = {
```
Comment (FinnStutzenstein) [inline client/src/app/site/users/services/group-repository.service.ts]: the `appId` must not be converted to string.

File: client/src/app/site/users/services/group-repository.service.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                        break;
+                    case 'users':
+                        this.addAppPerm(5, perm, 'Users');
+                        break;
+                    default:
+                        // plugins
+                        const displayName = `${permApp.charAt(0).toUpperCase}${permApp.slice(1)}`;
+                        // check if the plugin exists as app
+                        let result = -1;
+                        this.appPermissions.forEach((app, index) => {
```
Comment (FinnStutzenstein) [inline client/src/app/site/users/services/group-repository.service.ts]: Maybe use `findIndex` here. Should be shorter and more clear

File: client/src/app/site/users/components/group-list/group-list.component.ts [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+    /**
+     * Saves a newly created group.
+     * @param form form data given by the group
+     */
+    public submitNewGroup(form: any): void {
+        if (form.value) {
+            this.repo.create(form.value).subscribe(response => {
+                if (response) {
+                    form.reset();
```
[the remark this answers is not in the document]
Comment (tsiegleauq) [reply client/src/app/site/users/components/group-list/group-list.component.ts]: Should it really? Would be annoying to always click the button if you need to create more than one new group. I could close it i.e hitting the escape button.

[the remark this answers is not in the document]
Comment (FinnStutzenstein) [reply client/src/app/site/users/components/group-list/group-list.component.ts]: I was confused, that it wasn't closed. It's strange, if you edit a group afterwards and both forms are open. Jochen is OK with closing the form.

File: client/src/app/site/users/components/user-list/user-list.component.ts [hunk 4]
```diff
@@ -81,7 +81,7 @@ export class UserListComponent extends ListViewBaseComponent<ViewUser> implement
      * TODO: implement
      */
     public toGroups(): void {
-        console.log('to Groups');
+        this.router.navigate(['./groups'], { relativeTo: this.route });
```
[the remark this answers is not in the document]
Comment (FinnStutzenstein) [reply client/src/app/site/users/components/user-list/user-list.component.ts]: Note2: The baseperm thing does not work properly.. maybe skipping this is better ;)

[the remark this answers is not in the document]
Comment (tsiegleauq) [reply client/src/app/site/users/components/user-list/user-list.component.ts]: > the basePerm has to be set in the user routing module
Has extremely nasty side effects. Pressing F5 on every route that is protected in that way would bring you out to "/".
Would probably require to keep the operator in the localStorrage (again)

File: client/src/app/shared/models/users/group.ts [hunk 5]
```diff
@@ -11,6 +11,10 @@ export class Group extends BaseModel<Group> {
 
     public constructor(input?: any) {
         super('users/group', input);
+        if (!input) {
```
[the remark this answers is not in the document]
Comment (tsiegleauq) [reply client/src/app/shared/models/users/group.ts]: Constructing an object is essentially what a constructor is meant to do.
Furthermore, this changes seems to mess with the AutoUpdate routine.
The operator has no permissions anymore, so I gonna leave it as it is for now

File: client/src/app/site/users/components/group-list/group-list.component.html [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            <fa-icon icon="times"></fa-icon>
+        </button>
+    </form>
+</div>
+
+<mat-accordion *ngFor="let app of repo.appPermissions">
+    <mat-expansion-panel class="mat-elevation-z0" [expanded]=true>
+        <mat-expansion-panel-header>
+            <mat-panel-title>
+                {{app.name}}
```
[the remark this answers is not in the document]
Comment (tsiegleauq) [reply client/src/app/site/users/components/group-list/group-list.component.html]: added `translate` to the `mat-panel-title`

File: client/src/app/site/users/components/group-list/group-list.component.html [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+            </mat-panel-title>
+        </mat-expansion-panel-header>
+
+        <div class="scrollable-perm-matrix">
+            <table mat-table class='on-transition-fade' [dataSource]="getTableDataSource(app.permissions)">
+                <ng-container matColumnDef="perm" sticky>
+                    <mat-header-cell *matHeaderCellDef> Permissions </mat-header-cell>
+                    <mat-cell *matCellDef="let perm">
+                        {{perm.display_name}}
```
[the remark this answers is not in the document]
Comment (tsiegleauq) [reply client/src/app/site/users/components/group-list/group-list.component.html]: added `translate` to the `<mat-cell>`

File: client/src/app/site/users/components/group-list/group-list.component.html [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                    </mat-cell>
+                </ng-container>
+
+                <div *ngFor=" let group of groups; trackBy: trackGroupArray">
+                    <ng-container [matColumnDef]="group.name">
+                        <mat-header-cell class="group-head-table-cell" *matHeaderCellDef (click)="selectGroup(group)">
+                            <div class="inner-table">
+                                {{group.name}}
```
[the remark this answers is not in the document]
Comment (tsiegleauq) [reply client/src/app/site/users/components/group-list/group-list.component.html]: added `translate` to the `<div>`

File: client/src/app/site/users/components/group-list/group-list.component.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tsiegleauq) [reply client/src/app/site/users/components/group-list/group-list.component.ts]: Changed that for now. Also prevented the effect that both the form were shown.

**Comment by tsiegleauq** [comment]
Made the required changes (well, except the first one, since it would break stuff). Should be fine now (if travis permits...)

File: client/src/app/site/users/components/group-list/group-list.component.ts [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+/**
+ * Component for the Group-List and permission matrix
+ */
+@Component({
+    selector: 'os-group-list',
+    templateUrl: './group-list.component.html',
+    styleUrls: ['./group-list.component.scss']
+})
+export class GroupListComponent extends ListViewBaseComponent<any> implements OnInit {
```
Comment (FinnStutzenstein) [inline client/src/app/site/users/components/group-list/group-list.component.ts]: Is the inheritance really needed?

File: client/src/app/site/users/components/group-list/group-list.component.ts [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    public newGroupButton(): void {
+        this.editGroup = false;
+        this.newGroup = !this.newGroup;
+    }
+
+    /**
+     * Saves a newly created group.
+     * @param form form data given by the group
+     */
+    public submitNewGroup(form: any): void {
```
Comment (FinnStutzenstein) [inline client/src/app/site/users/components/group-list/group-list.component.ts]: type `FormGroup`

File: client/src/app/site/users/components/group-list/group-list.component.ts [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+                }
+            });
+        }
+    }
+
+    /**
+     * Saves an edited group.
+     * @param form form data given by the group
+     */
+    public submitEditedGroup(form: any): void {
```
Comment (FinnStutzenstein) [inline client/src/app/site/users/components/group-list/group-list.component.ts]: type `FormGroup`

**Pull request merged**